File: Grind75_1/week6/Subsets.py
"""
https://leetcode.com/problems/subsets/
"""

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        
        resList = [[]]

        lenNums = len(nums)

        inProc = [False] * lenNums

        def DFS(resList: List[List[int]], inProc: List[bool], tempList: List[int], idx: int):
            if (idx >= lenNums):
                return

            for i in range(idx, lenNums):
                if (inProc[i]):
                    continue
                else:
                    inProc[i] = True
                    tempList.append(nums[i])
                    resList.append(tempList[:])
                    DFS(resList, inProc, tempList, i + 1)
                    tempList.pop()
                    inProc[i] = False

            return


        DFS(resList, inProc, [], 0)

        return resList

File: neetcode/neet150_JS/Backtracking/subsets.js
// https://neetcode.io/problems/subsets

class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums) {
        if (nums.length === 0) {
            return []
        }

        const res = []
        const subset = []
        this.DFS(nums, 0, res, subset)
        return res
    }

    DFS(nums, i, res, subset) {
        if (i >= nums.length) {
            res.push(Array.from(subset))
            return
        }

        subset.push(nums[i])
        this.DFS(nums, i + 1, res, subset)

        subset.pop()
        this.DFS(nums, i + 1, res, subset)
    }
}
